RUMM-139 Add documentation for logging (#46)
DataDog/dd-sdk-ios: docs/log_collection.md

> # IOS Log Collection

Author: ```suggestion
# iOS Log Collection
```

> Send logs to Datadog from your IOS applications with [Datadog's `dd-sdk-ios` client-side logging library][1] and leverage the following features:

Author: ```suggestion
Send logs to Datadog from your iOS applications with [Datadog's `dd-sdk-ios` client-side logging library][1] and leverage the following features:
```

* Log to Datadog in JSON format natively.
* Add `context` and extra custom attributes to each log sent.
* Record real client IP addresses and User-Agents.
* Optimized network usage with automatic bulk posts.

> **Note**: The `dd-sdk-ios` library supports all IOS versions from 11.

Author: ```suggestion
**Note**: The `dd-sdk-ios` library supports all iOS versions 11+.
```

> **Note**: The `dd-sdk-ios` library supports all IOS versions from 11.

Reviewer: done

## Setup

1. Declare the library as a dependency depending on your package manager:

    {{< tabs >}}
    {{% tab "CocoaPods" %}}

You can use [CocoaPods][6] to install `dd-sdk-ios`
```
pod 'Datadog'
```

    {{% /tab %}}
    {{% tab "Swift Package Manager (SPM)" %}}

To integrate using Apple's Swift package manager, add the following as a dependency to your Package.swift:
```swift
.package(url: "https://github.com/Datadog/dd-sdk-ios.git", .upToNextMajor(from: "0.0.0"))
```

    {{% /tab %}}
    {{% tab "Carthage" %}}

> Not yet supported

Author: If there are no very immediate plans to support carthage, let's just remove this tab

    {{% /tab %}}
    {{< /tabs >}}

> 2. Initialize the library with your application context and your [Datadog client token][2]. For security reasons, you must use a client token: you cannot use [Datadog API keys][3] to configure the `dd-sdk-ios` library as they would be exposed client-side in the IOS application IPA byte code. For more information about setting up a client token, see the [client token documentation][2]:

Author: ```suggestion
2. Initialize the library with your application context and your [Datadog client token][2]. For security reasons, you must use a client token: you cannot use [Datadog API keys][3] to configure the `dd-sdk-ios` library as they would be exposed client-side in the iOS application IPA byte code. For more information about setting up a client token, see the [client token documentation][2]:
```

    {{< tabs >}}
    {{% tab "US" %}}

```swift
Datadog.initialize(
    appContext: .init(),
    configuration: Datadog.Configuration
        .builderUsing(clientToken: config.clientToken)
        .build()
)
```

    {{% /tab %}}
    {{% tab "EU" %}}

```swift
Datadog.initialize(
    appContext: .init(),
    configuration: Datadog.Configuration
        .builderUsing(clientToken: config.clientToken)
        .set(logsEndpoint: .eu)
        .build()
)
```

    {{% /tab %}}
    {{< /tabs >}}

> When writing your application, you can enable development logs. All internal messages in the library with a priority equal or higher than the provided level are then logged to console logs.

Author: ```suggestion
When writing your application, you can enable development logs. All internal messages in the library with a priority equal to or higher than the provided level are then logged to console logs.
```

```swift
    Datadog.verbosityLevel = .debug
```

> 3. Configure the IOS Logger:

Author: ```suggestion
3. Configure the iOS Logger:
```

    ```swift
    logger = Logger.builder
        .set(serviceName: "ios-sdk-example-app")
        .printLogsToConsole(true, usingFormat: .shortWith(prefix: "[iOS App] "))
        .build()
    ```

> 4. Send a custom log entry directly to Datadog with one of the following functions:

Reviewer 2: ```suggestion
4. Send a custom log entry directly to Datadog with one of the following methods:
```

    logger.info("Some relevant information?")
    logger.notice("Have you noticed?")
    logger.warn("An important warning…")
    logger.error("An error was met!")
    logger.critical("Something critical happened!")
    ```

5. (Optional) - Provide a map of `attributes` alongside your log message to add attributes to the emitted log. Each entry of the map is added as an attribute.

    ```swift
    logger.info("Clicked OK", attributes: ["context": "onboarding flow"])
    ```

## Advanced logging

### Initialization

The following methods in `Logger.Builder` can be used when initializing the logger to send logs to Datadog:

| Method                           | Description                                                                                                                                                                                                                         |
|----------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `sendNetworkInfo(_ enabled: true)`    | Add `network.client.connectivity` attribute to all log. The data logged by default is `connectivity` (`Wifi`, `3G`, `4G`...) and `carrier_name` (`AT&T - US`). |

Author: ```suggestion
| `sendNetworkInfo(_ enabled: true)`    | Add `network.client.connectivity` attribute to all logs. The data logged by default is `connectivity` (`Wifi`, `3G`, `4G`...) and `carrier_name` (`AT&T - US`). |
```

> | `set(serviceName: <SERVICE_NAME>)` | Set `<SERVICE_NAME>` as value for the `service` [standard attribute][4] attached to all logs sent to Datadog.                                                                                                                        |

Author: ```suggestion
| `set(serviceName: <SERVICE_NAME>)` | Set `<SERVICE_NAME>` as the value for the `service` [standard attribute][4] attached to all logs sent to Datadog.                                                                                                                        |
```

| `printLogsToConsole(_ enabled: true, usingFormat format: ConsoleLogFormat = .short)`     | Set to `true` to send logs to the debugger console.                                                                                                                                                                                         |
| `sendLogsToDatadog(_ enabled: true)`    | Set to `true` to send logs to Datadog.                                                                                                                                                                                              |
| `set(loggerName: <LOGGER_NAME>)`   | Set `<LOGGER_NAME>` as the value for the `logger.name` attribute attached to all logs sent to Datadog.                                                                                                                                   |
| `build()`                        | Build a new logger instance with all options set.                                                                                                                                                                                   |

### Global configuration

> Find below functions to add/remove tags and attributes to all logs sent by a given logger.

Reviewer 2: ```suggestion
Find below methods to add/remove tags and attributes to all logs sent by a given logger.
```

#### Global Tags

##### Add Tags

> Use the `addTag(withKey: "<TAG_KEY>", value: "<TAG_VALUE>")` function to add tags to all logs sent by a specific logger:

Reviewer 2: ```suggestion
Use the `addTag(withKey: "<TAG_KEY>", value: "<TAG_VALUE>")` method to add tags to all logs sent by a specific logger:
```

```swift
// This adds a tag "build_configuration:debug"
logger.addTag(withKey: "build_configuration", value: "debug")
```

**Note**: `<TAG_VALUE>` must be a String.

##### Remove Tags

> Use the `removeTag(withKey key: "<TAG_KEY>")` function to remove tags from all logs sent by a specific logger:

Reviewer 2: ```suggestion
Use the `removeTag(withKey key: "<TAG_KEY>")` method to remove tags from all logs sent by a specific logger:
```

```swift
// This removes any tag starting with "build_configuration"
logger.removeTag(withKey key: "build_configuration")
```

[Learn more about Datadog tags][5].

#### Global Attributes

##### Add attributes

By default, the following attributes are added to all logs sent by a logger:

* `http.useragent` and its extracted `device` and `OS` properties

> * `network.client.ip` and its extracted geographical properties (`country`, `city`)

Reviewer 2: Also:
```
"logger.version": "\(sdkVersion)", // Datadog SDK version
"logger.thread_name" : "main", // either "main" or "background"
"application.version": "1.0.0" // client's app version extracted from `Info.plist`
```

Reviewer: Done

> Use the `addAttribute(forKey key: "<ATTRIBUTE_KEY>", value: "<ATTRIBUTE_VALUE>")` function to add a custom attribute to all logs sent by a specific logger:

Reviewer 2: ```suggestion
Use the `addAttribute(forKey key: "<ATTRIBUTE_KEY>", value: "<ATTRIBUTE_VALUE>")` method to add a custom attribute to all logs sent by a specific logger:
```

```swift
// This adds an attribute "device-model" with a string value
logger.addAttribute(forKey: "device-model", value: UIDevice.current.model)
```

> **Note**: `<ATTRIBUTE_VALUE>` can be any primitive, String, or Date.

Reviewer 2: ```suggestion
**Note**: `<ATTRIBUTE_VALUE>` can be anything conforming to `Encodable` (String, Date, custom `Codable` data model, ...).
```

##### Remove attributes

> Use the `removeAttribute(forKey key: "<ATTRIBUTE_KEY>")` function to remove a custom attribute from all logs sent by a specific logger:

Reviewer 2: ```suggestion
Use the `removeAttribute(forKey key: "<ATTRIBUTE_KEY>")` method to remove a custom attribute from all logs sent by a specific logger:
```

# iOS Log Collection

Send logs to Datadog from your iOS applications with [Datadog's `dd-sdk-ios` client-side logging library][1] and leverage the following features:

* Log to Datadog in JSON format natively.
* Add `context` and extra custom attributes to each log sent.
* Record real client IP addresses and User-Agents.
* Optimized network usage with automatic bulk posts.

> **Note**: The `dd-sdk-ios` library supports all iOS versions 12+.

Reviewer 3: 11+ now right?

Reviewer 2: Ha! 🏅

> **Note**: The `dd-sdk-ios` library supports all iOS versions 12+.

Reviewer 2: ```suggestion
**Note**: The `dd-sdk-ios` library supports all iOS versions 11+.
```

## Setup

1. Declare the library as a dependency depending on your package manager:

    {{< tabs >}}
    {{% tab "CocoaPods" %}}

You can use [CocoaPods][6] to install `dd-sdk-ios`:
```

> pod 'Datadog'

Reviewer 2: ```suggestion
pod 'DatadogSDK'
```
